### ISSUE TITLE
No data received if one or more nodes not ready in kubernetes

### ISSUE THREAD

**Issue by FlavioF** [kubernetes/heapster#787]
Hi,

I am using Heapster with Kubernetes. My cluster has a size of 4 nodes.
When one of the nodes stops (and changed state to not ready), the Heapster stops to send data to InfluxDB.

In the example, I stopped the node 3 of my cluster.

Heapster is logging this message:

```
E1214 12:37:55.008564       1 kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
E1214 12:38:00.008963       1 kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
E1214 12:38:35.000742       1 kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
E1214 12:39:05.003094       1 kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
E1214 12:39:35.004320       1 kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
```

```
NAME                                        LABELS                                                             STATUS
node-1-samantha.internal   kubernetes.io/hostname=node-1-samantha.internal   Ready
node-2-samantha.internal   kubernetes.io/hostname=node-2-samantha.internal   Ready
node-3-samantha.internal   kubernetes.io/hostname=node-3-samantha.internal   NotReady
node-4-samantha.internal   kubernetes.io/hostname=node-4-samantha.internal   Ready
```

Data presented in Grafana:
<a href='http://postimg.org/image/u1plc0ejj/' target='_blank'><img src='http://s11.postimg.org/u1plc0ejj/Selection_012.jpg' border='0' alt="Selection 012" /></a>

Versions:

> Heapster Canary
> Kubernetes 1.0.1
> InfluxDB 0.9.4.1

**Comment by vishh** [comment]
That's not expected. Heapster will continue its operations even if it encounters any errors.
cc @mwielgus

**Comment by mwielgus** [comment]
1. Can you tell me the exact Heapster version (last git commit)?
2. Can you share some more details about your deployment? Is it Amazon, GCE or some bare-metal cluster.
3. Did Heapster stop writing the metrics completely? (No metrics for other pods as well)
4. Did the node come back to life? If yes, did Heapster recover?

**Comment by FlavioF** [comment]
1. I was using the canary docker image in docker hub, but now I changed the version to 0.19.0 and it is still happening
2. GCE using my own kubernetes cluster (not GKE)
3. Yes
4. Yes

Using 0.19.0 this log message doesn't appear but heapster didn't send any data when I shut down one of the nodes
```
kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"
```

**Comment by huangyuqi** [comment]
Hi, @FlavioF 

+ Can you request the Heapster's RESTApi to get the metrics? If you can get any data, maybe there is something wrong about the influxDB sink? You can change the log level to get more details.You can try this RESTapi: `http://***:8082/api/v1/metrics`

+ AFAIK, Heapster has aging mechanism; So if you stopped a node, it will exist in `node list` of cache ,until heapster `runGc` to delete. Thus, you will get the error log `kube_nodes.go:59] Failed to get container stats from Kubelet on node "node-3-samantha.internal"` until removing the node from cache.So i think we can ignore this error log in current scene.

**Comment by FlavioF** [comment]
Hi, @huangyuqi 

I have tried the Heapster's RESTApi: **http://.../api/v1/model/namespaces/default/pods**. In the same scenario described above.

Rest results with all nodes alive:
```
[
  {
   "name": "hazelcast-3.5-rk8es",
   "cpuUsage": 2,
   "memUsage": 305459200
  },
...
]
```

Rest results with one node down are always:
```
[
  {
   "name": "hazelcast-3.5-rk8es",
   "cpuUsage": 0,
   "memUsage": 0
  },
...
]
```
I waited like 1 hour and there is still any data, so I believe it is not related to cache.
Is it possible that Heapster isn't removing the NotReady nodes from the list?

Honestly, Idk whats happening here. Is it only happening to me?

**Comment by huangyuqi** [comment]
Yes, @FlavioF , I have two nodes of kubernetes cluster,

```
8.8.8.237
8.8.8.239
```
when I stop one node `8.8.8.239` , i got the same issue:
`E0106 23:57:50.001755   26308 kube_nodes.go:59] Failed to get container stats from Kubelet on node "8.8.8.239"`

i have debugged this, i got this:
In heapster node function `List` and `getNodeInfoAndHostname`, heapster doesn't check the node state.  
https://github.com/kubernetes/heapster/blob/master/sources/nodes/kube.go#L110-L121

And, the checking process just takes `LookupIP `

```
	if nodeInfo.InternalIP == "" {
		if hostname == nodeInfo.PublicIP {
			// If the only identifier we have for the node is a public IP, then use it;
			// don't force a DNS lookup
			glog.V(4).Infof("Only have PublicIP %s for node %s, so using it for InternalIP",
				nodeInfo.PublicIP, node.Name)
			nodeInfo.InternalIP = nodeInfo.PublicIP
		} else {
			addrs, err := net.LookupIP(hostname)
			if err == nil {
				nodeInfo.InternalIP = addrs[0].String()
			} else {
				glog.Errorf("Skipping host %s since looking up its IP failed - %s", node.Name, err)
				self.recordNodeError(node.Name)
				nodeErr = err
			}
		}
	}

```

Even though the node is `not ready`, the heapster's nodeList also contains it. 
So i think this is a bug, I will take a commit `add node's state checking`to fix. @vishh  @mwielgus

**Comment by huangyuqi** [comment]
Hi, @FlavioF 
I have push a commit to fix this, please check https://github.com/kubernetes/heapster/pull/862, and debug in your enviroment. Thanks

**Comment by piosz** [comment]
fixed in #862
